modules/tinymce/src/core/main/ts/newline/InsertBlock.ts

```diff
@@ -242,139 +224,198 @@ const addBrToBlockIfNeeded = function (dom, block) {
+  if (NodeType.isText(container)) {
+    if (start && normalizedOffset === 0) {
+      walker.prev();
+    } else if (!start && normalizedOffset === container.nodeValue.length) {
+      walker.next();
     }
+  }
 
-    while ((node = walker.current())) {
-      if (NodeType.isElement(node)) {
-        // Ignore bogus elements
-        if (!node.getAttribute('data-mce-bogus')) {
-          // Keep empty elements like <img /> <input /> but not trailing br:s like <p>text|<br></p>
-          name = node.nodeName.toLowerCase();
-          if (nonEmptyElementsMap[name] && name !== 'br') {
-            return false;
-          }
+  let node: Node;
+  while ((node = walker.current())) {
+    if (NodeType.isElement(node)) {
+      // Ignore bogus elements
+      if (!node.getAttribute('data-mce-bogus')) {
+        // Keep empty elements like <img /> <input /> but not trailing br:s like <p>text|<br></p>
+        const name = node.nodeName.toLowerCase();
+        if (nonEmptyElementsMap[name] && name !== 'br') {
+          return false;
         }
-      } else if (NodeType.isText(node) && !isWhitespaceText(node.nodeValue)) {
-        return false;
-      }
-
-      if (start) {
-        walker.prev();
-      } else {
-        walker.next();
       }
+    } else if (NodeType.isText(node) && !isWhitespaceText(node.nodeValue)) {
+      return false;
     }
 
-    return true;
-  };
-
-  const insertNewBlockAfter = function () {
-    // If the caret is at the end of a header we produce a P tag after it similar to Word unless we are in a hgroup
-    if (/^(H[1-6]|PRE|FIGURE)$/.test(parentBlockName) && containerBlockName !== 'HGROUP') {
-      newBlock = createNewBlock(newBlockName);
+    if (start) {
+      walker.prev();
     } else {
-      newBlock = createNewBlock();
+      walker.next();
     }
+  }
 
-    // Split the current container block element if enter is pressed inside an empty inner block element
-    if (Settings.shouldEndContainerOnEmptyBlock(editor) && canSplitBlock(dom, containerBlock) && dom.isEmpty(parentBlock)) {
-      // Split container block for example a BLOCKQUOTE at the current blockParent location for example a P
-      newBlock = dom.split(containerBlock, parentBlock);
-    } else {
-      dom.insertAfter(newBlock, parentBlock);
-    }
+  return true;
+};
 
-    NewLineUtils.moveToCaretPosition(editor, newBlock);
+// Used as an internal function below. Dragging it up here to make the function state explicit (and read-only).
+interface InsertNewBlockAfter {
+  containerBlockName: string;
+  newBlockName: string;
+  parentBlockName: string;
+  createNewBlock: (name?: string) => Node;
+  parentBlock: Element;
+  containerBlock: Element;
+  dom: DOMUtils;
+  editor: Editor;
+}
+
+const _insertNewBlockAfter = ({ containerBlockName, newBlockName, parentBlockName, createNewBlock, parentBlock, containerBlock, dom, editor }: InsertNewBlockAfter): Node => {
```

Comment:
One more using underscore and destructing, which doesn't follow our normal conventions.